core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala

```diff
@@ -209,69 +281,34 @@ class QualificationAppInfo(
-            val singleNextStage = next.stages.headOption.toSeq
-            val dedupedStages = Set(singlePrevStage ++ singleNextStage ++ prevStages ++ nextStages).
-              flatten.toSeq
-            if (dedupedStages.nonEmpty) {
-              dedupedStages
-            } else {
-              // we don't know what stage its in or its duration
-              logDebug(s"No stage associated with ${execInfo.exec} " +
-                s"so speedup factor isn't applied anywhere.")
-              execsWithoutStages += execInfo
-              Seq.empty
-            }
-          } else {
-            Seq(execInfo.stages.head)
-          }
-        associatedStages.map((_, execInfo))
-
-      case Seq(prev, execInfo) =>
-        val associatedStages =
-          if (execInfo.stages.size > 1) execInfo.stages.toSeq
-          else if (execInfo.stages.size < 1) {
-            val prevStages = if (prev.stages.size > 1) prev.stages.toSeq else Seq.empty
-            val singlePrevStage = prev.stages.headOption.toSeq
-            val dedupedStages = Set(singlePrevStage ++ prevStages).flatten.toSeq
-            if (dedupedStages.nonEmpty) {
-              dedupedStages
-            } else {
-              // we don't know what stage its in or its duration
-              logDebug(s"No stage associated with ${execInfo.exec} " +
-                s"so speedup factor isn't applied anywhere.")
-              execsWithoutStages += execInfo
-              Seq.empty
-            }
-          } else {
-            Seq(execInfo.stages.head)
-          }
-        associatedStages.map((_, execInfo))
-
-      case Seq(execInfo) =>
-        val associatedStages =
-          if (execInfo.stages.size > 1) execInfo.stages.toSeq
-          else if (execInfo.stages.size < 1) {
-            execsWithoutStages += execInfo
-            Seq.empty
-          } else {
-            Seq(execInfo.stages.head)
-          }
-        associatedStages.map((_, execInfo))
-
-      case _ => Seq.empty
-    }.toSeq.flatten.groupBy(_._1).map { case (stage, execInfos) =>
-      (stage, execInfos.map(_._2))
+    val execsToStageMap = execInfosInOrder.indices.map {
+      // corner case to handle first element
+      case 0 => if (execInfosInOrder.size > 1) {
+        // If there are more than one Execs, then check if the next Exec has a stageId.
+        doubleExecMatch(execInfosInOrder(1), execInfosInOrder(0))
+      } else {
+        singleExecMatch(execInfosInOrder(0))
+      }
+      // corner case to handle last element
+      case i if i == execInfosInOrder.size - 1 =>
+        if (execInfosInOrder.size > 1) {
+          // If there are more than one Execs, then check if the previous Exec has a stageId.
+          doubleExecMatch(execInfosInOrder(i - 1), execInfosInOrder(i))
+        } else {
+          singleExecMatch(execInfosInOrder(i))
+        }
+      case i =>
+        tripleExecMatch(execInfosInOrder(i - 1), execInfosInOrder(i), execInfosInOrder(i + 1))
+      case _ =>
```

Comment:
We should not require this case statement as the above case statement would cover all scenarios

Reply:
updated.